modal/runner/engines/vllm.py

```diff
@@ -70,70 +82,56 @@ def __init__(self, params: VllmParams):
     async def generate(self, payload: CompletionPayload, params):
         assert self.engine is not None, "Engine not initialized"
 
+        # Track usage as a running total
+        # NOTE: This does NOT yet include cold-start GPU time
         t_start_inference = time.perf_counter()
+        usage = Usage(
+            prompt_tokens=0,
+            completion_tokens=0,
+            duration=0.0,
+            gpu_type=self.gpu_type,
+            gpu_count=self.gpu_count,
+        )
 
         try:
             results_generator = self.engine.generate(
                 payload.prompt, params, payload.id
             )
 
-            final_output = None
-            if payload.stream:
-                index = 0
-
-                async for request_output in results_generator:
-                    # Skipping invalid UTF8 tokens:
-                    if (
-                        request_output.outputs[0].text
-                        and request_output.outputs[0].text[-1] == "\ufffd"
-                    ):
-                        continue
-
-                    final_output = request_output
-                    token = final_output.outputs[0].text[index:]
-                    index = len(final_output.outputs[0].text)
-                    yield create_sse_data(
-                        token,
-                        prompt_tokens=len(final_output.prompt_token_ids),
-                        completion_tokens=len(
-                            final_output.outputs[0].token_ids
-                        ),
-                        done=False,
-                    )
-
-                output = ""
-            else:
-                async for request_output in results_generator:
-                    final_output = request_output
-                    yield " "
-
-                output = final_output.outputs[0].text
+            output = ""
+            index = 0
+            async for current in results_generator:
```

Comment:
golfing this function down so we have less nested `if/else` blocks & its easier to see what's shared by streaming & non-streaming request paths

will need to test a bit rigorously on dev first, but this block is where a more thorough review could be appreciated